scipy/stats/_multicomp.py

```diff
@@ -24,6 +26,133 @@
 class DunnettResult:
     statistic: np.ndarray
     pvalue: np.ndarray
+    _alternative: Literal['two-sided', 'less', 'greater']
+    _rho: np.ndarray
+    _df: int
+    _std: float
+    _mean_samples: np.ndarray
+    _mean_control: np.ndarray
+    _ci: ConfidenceInterval | None = None
+    _ci_cl: DecimalNumber | None = None
+
+    def __str__(self):
+        # Note: `__str__` prints the confidence intervals from the most
+        # recent call to `confidence_interval`. If it has not been called,
+        # it will be called with the default CL of .95.
+        if self._ci is None:
+            self.confidence_interval(confidence_level=.95)
+        s = (
+            "Dunnett's test"
+            f" ({self._ci_cl*100:.1f}% Confidence Interval)\n"
+            "Comparison               Statistic  p-value  Lower CI  Upper CI\n"
+        )
+        for i in range(self.pvalue.shape[0]):
+            s += (f" (Sample {i} - Control) {self.statistic[i]:>10.3f}"
+                  f"{self.pvalue[i]:>10.3f}"
+                  f"{self._ci.low[i]:>10.3f}"
+                  f"{self._ci.high[i]:>10.3f}\n")
+
+        if self._alternative == 'less':
+            s += (
+                "\nOne-sided alternative (less): sample i's mean exceed "
+                "the control's mean by an amount at least Lower CI"
+            )
+        elif self._alternative == 'greater':
+            s += (
+                "\nOne-sided alternative (greater): sample i's mean exceed "
+                "the control's mean by an amount at most Upper CI"
+            )
+        else:
+            s += (
+                "\nTwo-sided alternative: sample i's mean exceed the "
+                "control's mean by an amount between Lower CI and Upper CI"
+            )
+        return s
+
+    def _allowance(self, confidence_level: DecimalNumber = 0.95) -> float:
+        """Allowance.
+
+        It is the quantity to add/substract from the observed difference
+        between the means of observed groups and the mean of the control
+        group. The result gives confidence limits.
+
+        Parameters
+        ----------
+        confidence_level : float, optional
+            Confidence level for the computed confidence interval
+            of the estimated proportion. Default is .95.
+
+        Returns
+        -------
+        allowance : float
+            Allowance around the mean.
+        """
+        alpha = 1 - confidence_level
+
+        def pvalue_from_stat(statistic):
+            # two-sided to have +- bounds
+            statistic = np.array(statistic)
+            sf = _pvalue_dunnett(
```

Comment:
Because this is calculated with RQMC, this is a noisy function that will be tricky to optimize over. [`mvtnorm`'s `qmvt`](https://www.rdocumentation.org/packages/mvtnorm/versions/1.1-3/topics/qmvt) uses [a procedure that builds a small regression model](https://link.springer.com/article/10.1007/s00180-017-0736-0) from the noisy samples as it does the root-finding.

Reply:
I don't understand what you are suggesting.

Reply:
I'm suggesting that the method in the linked paper is the best way to implement this root-finding. This current implementation will run into problems sooner or later, if not now.

Reply:
mmm but there is just a single citation on the linked paper.

Reply:
That's fine by me. It was developed for `mvtnorm` and so many of the [~6000 citations for `mvtnorm`'s preferred citation since the algorithm's inclusion](https://scholar.google.com/scholar?as_ylo=2019&hl=en&as_sdt=20000005&sciodt=0,21&cites=13733321422736086043&scipsc=) would count, IMO, if citation count is our criteria.

Reply:
> This current implementation will run into problems sooner or later, if not now.

Do we actually have any evidence of that? Can we construct a realistic example that would fail?

What worries me is that, we would be using a new root finding method as opposed to a battle tested method that we already have?? That sounds less robust to me.

Reply:
I did a small study using the univariate t distribution as a proxy for the multivariate t, which may or may not be representative.

In short, I used `minimize_scalar` to minimize `abs(noisy_cdf(x) - p)`, where `noisy_cdf(x) = stats.t(df).cdf(x) + rng.normal(loc=0, scale=noise_scale)` and `p` is the target value of the CDF. I was interested in how the `noise_scale` affects the error in the solution.

<details>

```
import numpy as np
from scipy import stats, optimize
import matplotlib.pyplot as plt
from mpmath import mp
mp.dps = 50
rng = np.random.default_rng(3882823079058616272161)

# Define noisy CDF
dist = stats.t(12)
noise_scale = 1e-4
def noisy_cdf(x):
    return dist.cdf(x) + rng.normal(loc=0, scale=noise_scale)

# Numerically invert many times
p = 0.05
xs = []
x0 = dist.ppf(p)

for i in range(10000):
    res = optimize.minimize_scalar(lambda x: abs(noisy_cdf(x) - p))
    xs.append(res.x)

# Look at the error in the result
plt.hist(xs, bins=30)
plt.title(f'Inverse of Noisy CDF, `p={p}`, `noise_scale`={noise_scale}')
plt.xlabel('res.x')
plt.ylabel('frequency')
plt.show()

print(x0)
print((max(xs) - x0)/x0, (min(xs) - x0)/x0)  # max/min error
```

</details>

When the `noise_scale` is small (`noise_scale=1e-4` here), there is usually little error in the output, although there were a few cases in which the error was as much as 5%.

![image](https://user-images.githubusercontent.com/6570539/225720161-e1898158-7a20-456d-92bf-a57ca56ee200.png)

When the `noise_scale` is larger (`noise_scale=1e-3` here), problems occur more frequently.

![image](https://user-images.githubusercontent.com/6570539/225720310-b01d6d85-567f-403c-85ab-3c49934ef0a5.png)


With larger `noise_scale` (`noise_scale=1e-2`), we can get complete breakdown.

![image](https://user-images.githubusercontent.com/6570539/225720491-cb3582d2-e90f-433a-85a4-2d87ad9414fc.png)

When `p` is very close to zero, the problem is even more severe for a fixed noise level. The error does not seem to be affected much by the requested tolerance or the method (e.g. `golden`). It might be improved by defining an initial bracket or imposing `bounds`.

So @rkern is right that we could have problems if there is too much noise.

---

Then there is the question of how much noise we have. For a balanced design (4 observations per group) with three experimental groups and a moderate value of the statistic:

<details>

```python3
from scipy.stats._multicomp import  _pvalue_dunnett, _params_dunnett
from scipy import stats

samples = [
      [55, 64, 64],
      [55, 49, 52],
      [50, 44, 41]
  ]
control = [55, 47, 48]
confidence_level = 0.95
alternative = 'two-sided'

samples, control = np.atleast_1d(samples, control)
rho, df, n_group = _params_dunnett(samples=samples, control=control)

alpha = 1 - confidence_level
def pvalue_from_stat(statistic):
    # two-sided to have +- bounds
    statistic = np.array(statistic)
    sf = _pvalue_dunnett(rho=rho, df=df, statistic=statistic,
                          alternative=alternative)
    return sf

ps = []
statistic = -2
for i in range(10000):
    ps.append(pvalue_from_stat(statistic))

# Look at the error in the result
plt.hist(ps, bins=20)
plt.title(f'p-value of dunnett test, statistic={statistic}')
plt.xlabel('p')
plt.ylabel('frequency')
plt.show()

print(np.std(ps, ddof=1))  # 4.378691608149277e-05
```

</details>

![image](https://user-images.githubusercontent.com/6570539/225729262-0e6e625f-d2c1-404b-b66e-177759770527.png)

The noise is quite small. For different values of the statistic, I didn't see noise scales (standard deviation) much greater than `1e-4`. The noise profiles looked either normal or skew-normal.... like everything else. But they weren't bimodal or anything weird.

For the highly unbalanced Matlab example, the noise actually tended to be smaller.

![image](https://user-images.githubusercontent.com/6570539/225728923-eb414361-af72-40bb-af4f-86af720853ae.png)

So as @tupui says, we might be able to get away with using `minimize_scalar` here. For these problems, at least, noise levels on the order of `1e-5` were typical. 


---

Combining the two parts on the multivariate t CDF is more time-consuming, but for the small, balanced design example above, the spread in the allowance calculation with `confidence_level=0.95` looks like:

![image](https://user-images.githubusercontent.com/6570539/225732981-931619fe-d0c3-4c1a-b0ff-5b35be8663b3.png)

With `confidence_level=0.99`:
![image](https://user-images.githubusercontent.com/6570539/225733398-5f96435d-3c64-415f-851f-41955a3e362a.png)

For the unbalanced Matlab example with `confidence_level=0.99`:
![image](https://user-images.githubusercontent.com/6570539/225734399-ee5af25d-3bb1-4d86-9b11-4d4188447660.png)

---

If we don't find reasonable counterexamples in the course of testing, I'd be comfortable working with `minimize_scalar` - perhaps with some tweaks on the termination condition - in this initial PR. Maybe we can verify that the solution returned by `minimize_scalar` is accurate by evaluating the CDF a few times at that statistic. If it varies too much, we can raise an error or return NaN so at least we don't give a wrong answer.

So with a little more testing and tweaking, I think this could be useful as-is, and we could add an improved method as a followup.

Reply:
`minimize_scalar(method='brent')` is definitely _not_ battle-tested for noisy functions like this. The noisiness violates the smoothness assumptions. You can see [some discussion here](https://www.johndcook.com/blog/2012/06/14/root-finding-with-noisy-functions/) (note the comments for the specific discussion of Brent's method). Brent's method for root-finding is related to the Brent's method for optimization here. Algorithms for root-finding or minimization on smooth functions need to be at least adapted to work well in noisy cases. They may work okay out of box, but that is by no means assured by their track record on the noiseless smooth functions that they were designed for. Applying them here is an ad hoc procedure that does not have any particular track record.

`mvtnorm` went through several iterations of quantile estimation before settling on the one I cited. First, they just used a Brent root-finder (i.e. finding where `p_value(x) - p == 0`) then moved to minimizing the squared error (i.e. minimizing `(p_value(x) - p)**2`), which is close to what you've implemented here. Just several months later, [they replaced it with the stochastic root-finding algorithm](https://github.com/cran/mvtnorm/blob/master/inst/NEWS#L68-L76) that is currently in use.

I do take that as evidence of a problem that they avoided by moving to the stochastic root finder. This is [part of a long tradition](https://en.wikipedia.org/wiki/Stochastic_approximation) of methods that deal with noisy functions like the one that we have. This is just the application of that theory to our specific problem. It's the workhorse algorithm for computing confidence intervals in `multcomp`, with thousands of citations after the algorithm's introduction. _This_ is the algorithm that is the most battle-tested for our problem, here.

I'm not opposed to the current approach as a stopgap, easily-available approach. Sometimes, they do work okay out of box. We should just understand that it is an ad hoc method and not a battle-tested, known robust one. Our default settings for the integrator tend to give us a noise about `1e-5`, which is pretty low. If you want to stress test things, higher dimensions (like 100) and higher confidence values like `0.99` will put you into a flatter region of the function. A correlation matrix with a value of `0.5` on each off-diagonal element is a typical one (it's what you get when you have balanced design in the Dunnett test). Probably the best bang for buck will be using better initial bounds rather than using the default bracket-finding.

Reply:
Thank you for the analysis @mdhaber! I will see to work with your suggestion to compute the value a few time and raise a warning.

@rkern by battle tested I was not clear. I am talking about the actual implementation in SciPy. Not about the sound mathematical truthiness. Meaning that if we were to add another optimiser, we would first add a maintenance component and second the uncertainty to have a robust software implementation. Hence if the current approach is viable-and from Matt is saying, I my understanding is that it is the case-, I personally would be more confident going with it. (Unless I missed the fact that we have a similar method already in SciPy??)

Reply:
I reviewed the paper, and don't think we do - at least not in those terms. 
@rkern Can a weighted generalized linear model with probit link be fit with weighted linear least squares? If so, perhaps the implementation is not so difficult, but I don't think the increase in scope is negligible. It would be better left to a separate PR, IMO.